[FEATURE] Feature/prepare for training feedbacktask

davidberenstein1957 opened the pull request:
# Description

I added a very rough outline of my ideation behind `prepare_for_training` with the new `FeedbackDataset`. As discussed there are 3 complexities:

- How to resolve annotator alignment?
- How to resolve optional fields, which have not been filled out? e.g., "Please provide a correction for prompt 1?".
- How handle potential concatenation of fields? 

To make it modular I created a step-wise approach.

1. `Pydantic` Models that map and verify data fields, like so. By doing this we keep the flexibility to allow for other tasks like TextClassification and this ensures we can directly use `datasets.field` and `dataset.questions` for defining training. We could also use the `name` values from the fields/questions, but this might be more error prone.
2. `get_relevant_data_for_training()` in `List[dict]` format with all relevant fields from the Pydantic model. **annotator alignment issue**. For now I opted for choosing the first non-zero value.
3. Forward the `List[dict]` to a similar flow we previously had.
4. Also add `dataset.unify_responses(question, Enum(strategy))`-method
5. Added `*QuestionUnifcation` to schemas to hold logic surrounding unifying multiplier responses
6. Added `client.feedback.training`
7. Added`TrainingDataFor*` to hold logic surrounding `prepare_for_training`-methods per task
8. Added inheritance for ArgillaTrainer

```python 
import argilla as rg
from argilla import (
    FeedbackRecord,
    LabelQuestion,
    LabelQuestionUnification,
    MultiLabelQuestion,
    TrainingDataForTextClassification,
    ArgillaTrainer
)

dataset = rg.FeedbackDataset(
    guidelines="Add some guidelines for the annotation team here.",
    fields=[
        rg.TextField(name="text", title="Human prompt"),
    ],
    questions =[
        LabelQuestion(
            name="relevant",
            title="Is the response relevant for the given prompt?",
            labels=["yes","no"],
            required=True,
            visible_labels=None
        ),
        MultiLabelQuestion(
            name="content_class",
            title="Does the response include any of the following?",
            description="Select all that apply",
            labels={"hate": "Hate Speech" , "sexual": "Sexual content", "violent": "Violent content", "pii": "Personal information", "untruthful": "Untruthful info", "not_english": "Not English", "inappropriate": "Inappropriate content"},
            required=False,
            visible_labels=4
        ),
    ]
)
dataset.add_records(
    records=[
        FeedbackRecord(
            fields={"text": "What is your favorite color?"},
            responses=[{"values": {"relevant": {"value": "yes"}, "content_class": {"value": ["hate"]}}}]
        ),
        FeedbackRecord(
            fields={"text": "What do you think about the new iPhone?"},
            responses=[{"values": {"relevant": {"value": "no"}, "content_class": {"value": ["hate"]}}}]
        ),
        FeedbackRecord(
            fields={"text": "What is your feeling about the technology?"},
            responses=[{"values": {"relevant": {"value": "yes"}, "content_class": {"value": ["sexual"]}}},
                       {"values": {"relevant": {"value": "no"}, "content_class": {"value": ["hate", "sexual"]}}},
                       {"values": {"relevant": {"value": "yes"}, "content_class": {"value": ["hate", "sexual"]}}}]
        ),
        FeedbackRecord(
            fields={"text": "Jesus Christ!"},
            responses=[{"values": {"relevant": {"value": "no"}, "content_class": {"value": ["sexual"]}}},
                       {"values": {"relevant": {"value": "no"}, "content_class": {"value": ["hate"]}}}]
        )

    ]
)

# print(dataset.question_by_name("relevant").__all_labels__)

label = LabelQuestionUnification(question=dataset.question_by_name("relevant"), strategy="majority")
training_data = TrainingDataForTextClassification(text=dataset.field_by_name("text"), label=label)

for framework in ["spacy", "transformers", "openai", "spark-nlp"]:
    formatted_data = dataset.prepare_for_training(framework, training_data, fetch_records=False, train_size=0.8)
    print(formatted_data)

trainer = ArgillaTrainer(
    dataset=dataset,
    training_task_mapping=training_task_mapping,
    framework="setfit",
    fetch_records=False
)
trainer.train("test")
```

Closes #2954
Closes #3184
Closes #3152 

**Type of change**

- [X] New feature (non-breaking change which adds functionality)
- [X] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

- [ ] Test A
- [ ] Test B

**Checklist**

- [ ] I have merged the original branch into my forked branch
- [ ] I added relevant documentation
- [ ] follows the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

davidberenstein1957 committed:
chore: added schemas for `RatingQuestionStrategy`
chore: added schemas for `LabelQuestionStrategy`
chore: added schemas for `MultiLabelQuestionStrategy`
chore: added schemas for `TrainingDataForTextClassification`

davidberenstein1957 committed:
chore: added unify_responses methods to strategies
chore: added unify_responses method to dataset

davidberenstein1957 committed:
chore: added working setting and strategy initialization

davidberenstein1957 committed:
chore: added basic unification strategies

davidberenstein1957 committed:
Merge branch 'develop' into feature/prepare-for-training-feedbacktask

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

davidberenstein1957 committed:
chore: added field_by_name-method

davidberenstein1957 committed:
chore: added unified response values to the main FeedbackRecord

davidberenstein1957 committed:
chore:  updated to in-memory unified_responses

davidberenstein1957 committed:
chore: added class inheritance of ValueSchema to UnificationValueSchema

frascuchon commented on src/argilla/client/feedback/schemas.py:
```diff
@@ -452,3 +475,253 @@ class FeedbackDatasetConfig(BaseModel):
+
+    text: TextField
+    label: Union[str, RatingQuestion, LabelQuestion, MultiLabelQuestion]
+    label_strategy: Union[RatingQuestionStrategy, LabelQuestionStrategy] = None
```
We can group label and label strategy in a BaseModel as :
```python
    class Rating(BaseModel):
        question: RatingQuestion
        strategy: RatingQuestionStrategy = RatingQuestionStrategy.MEAN

    class Label(BaseModel):
        question: Union[LabelQuestion, MultiLabelQuestion]
        strategy: LabelQuestionStrategy = LabelQuestionStrategy.MAJORITY

The, we can use them in the training model definition and reduce the root validator logic:
```suggestion
    text: TextField
    label: Union[Rating, Label]
```

davidberenstein1957 committed:
chore: added unified pydantic model group

davidberenstein1957 committed:
chore: initial version `prepare_for_training` for all frameworks

davidberenstein1957 committed:
chore: started integrationon with trainer frameworks

davidberenstein1957 committed:
chore: inherited behavior for new version of trainers

davidberenstein1957 committed:
chore: updated naming *_by_name-methods

davidberenstein1957 committed:
chore: inherit trainer v1

davidberenstein1957 committed:
fix: resolved circular import FeedbackDataset

davidberenstein1957 committed:
fix: resolved abs(None) execution

davidberenstein1957 committed:
chore: added strategy tests

davidberenstein1957 committed:
chore: added unificationstrategy tests

davidberenstein1957 committed:
fix: resolved None comparison test data

davidberenstein1957 committed:
fix: resolved failing span_marker tests

davidberenstein1957 committed:
chore: added training schema tests

davidberenstein1957 committed:
fix: added tests for trainer for FeedbackTask

davidberenstein1957 committed:
fix: added test training moduels prepare_for_training

davidberenstein1957 committed:
fix: resolved failing tests
chore: removed support autotrain

davidberenstein1957 committed:
chore: updated docs

davidberenstein1957 committed:
Merge branch 'develop' into feature/prepare-for-training-feedbacktask

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

davidberenstein1957 committed:
fix: resolved failing tests

davidberenstein1957 committed:
fix: added bugfixes usage peft trainer

davidberenstein1957 committed:
docs: updated documentation

davidberenstein1957 committed:
docs: reformatted some text

davidberenstein1957 committed:
docs: added warning about HF spaces
docs: added more references to training

alvarobartt commented on README.md:
```diff
@@ -85,7 +85,7 @@ docker run -d --name argilla -p 6900:6900 argilla/argilla-quickstart:latest
 <hr>
 <h3><a href="https://argilla.io/blog/launching-argilla-huggingface-hub/">Deploy on Hugging Face Hub</a></h3>
-
+WARNING: Hugging Face Spacese spaces don't have data retention by default and reset after 48 hours but this can be enabled within the settings on Hugging Face.
```
Typo-fix and removing the last part not to spoil anyone as not announced yet :/

```suggestion
WARNING: HuggingFace Spaces don't have persistent storage and will automatically reset after 48 hours of inactivity, which means all the data within the Space will be lost.
```

davidberenstein1957 replied:
I missed "private" beta

alvarobartt commented on docs/_source/_common/snippets/training/feedback-task/text-classification.md:
```diff
@@ -0,0 +1,33 @@
+
+dataset = rg.FeedbackDataset.from_argilla("<my_dataset_name>", workspace="<my_workspace_name>")
+
+label_unification = LabelUnification(
```
```suggestion
label_unification = rg.LabelUnification(
```

alvarobartt commented on docs/_source/_common/tabs/argilla_install.md:
````diff
@@ -15,6 +15,9 @@ wget -O docker-compose.yaml https://raw.githubusercontent.com/argilla-io/argilla
 
 :::{tab-item} Hugging Face Spaces
+```{warning}
+Hugging Face Spacese spaces don't have data retention by default so but this can be enabled within the spaces settings on Hugging Face.
````
Same here

```suggestion
HuggingFace Spaces don't have persistent storage and will automatically reset after 48 hours of inactivity, which means all the data within the Space will be lost.
```

alvarobartt commented on docs/_source/_common/tabs/unfication_strategies.md:
```diff
@@ -0,0 +1,55 @@
+Use a `Feedbackdataset` directly:
```
```suggestion
Use a `FeedbackDataset` directly:
```

alvarobartt commented on docs/_source/_common/tabs/unfication_strategies.md:
````diff
@@ -0,0 +1,55 @@
+```python
+from argilla import FeedbackDataset
+
+dataset = Feedbackdataset(...)
````
If initialized like this we won't have any response, so feel free to re-direct users either to an existing Argilla dataset in HugginFace via `from_huggingface` or to a placeholder with either `from_huggingface` and/or `from_argilla` instead

davidberenstein1957 replied:
We discussed this today, will see what I can do.

alvarobartt commented on docs/_source/_common/tabs/unfication_strategies.md:
````diff
@@ -0,0 +1,55 @@
+dataset.records[0].unified_response
+```
+
+Or, you can use `FeedbackRecord`s in combination with a `*QuestionStrategy`.
````
```suggestion
Or, you can use `FeedbackRecord`s in combination with a `QuestionStrategy`.
```

alvarobartt commented on docs/_source/getting_started/cheatsheet.md:
```diff
@@ -217,7 +217,7 @@ rg.log(records_for_training, name="majority_voter_results")
 ## Train Models
 
-We love our open-source training libraries as much as you do, so we provide integrations with all of them to limit the time you spend on data preparation and have more fun with actual training. As of now, we support `spacy`, `transformers`, `setfit`, `openai`, `autotrain`, and way more. Want to get to know all support? Take a look [here](/guides/train_a_model.md).
+We love our open-source training libraries as much as you do, so we provide integrations with all of them to limit the time you spend on data preparation and have more fun with actual training. As of now, we support `spacy`, `transformers`, `setfit`, `openai`, `autotrain`, and way more. Want to get to know all support? For the `FeedbackDataset`: take a look [here](/guides/train_a_model.md). For t ...
```
```suggestion
We love our open-source training libraries as much as you do, so we provide integrations with all of them to limit the time you spend on data preparation and have more fun with actual training. As of now, we support `spacy`, `transformers`, `setfit`, `openai`, `autotrain`, and way more. Want to get to know all support? Train/fine-tune a model from a `FeedbackDataset` as explained [here](/guides/train_a_model.md), or either a `TextClassificationDataset` or a `TokenClassificationDataset`[here](/guides/llms/fine_tune_others.md).
```

alvarobartt commented on docs/_source/getting_started/installation/configurations/user_management.md:
```diff
@@ -62,7 +62,7 @@ User with default credentials succesfully created:
 
 :::{note}
-To connect to an old Argilla instance (`<1.3.0`) using newer clients, you should specify the default user API key `rubrix.apikey`.
+To connect to an old Argilla instance (`<1.3.0`) using newer clients, you should specify the default user API key `rubrix.apikey` and for the quickstart image the default user API key is `admin.apikey`.
```
Isn't the default API Key now `argilla.apikey`? I mean as of 1.11.0 with the new `owner` role

```suggestion
To connect to an old Argilla instance (`<1.3.0`) using newer clients, you should specify the default user API key `rubrix.apikey`, for greater versions until `1.10.0` you should use the default API key `admin.apikey`, and from `1.11.0` onwards you should use `argilla.apikey`.
```

davidberenstein1957 replied:
@alvarobartt This was about the Quickstart image. Did we change that to `argilla.apikey`too?

alvarobartt replied:
Indeed we're still using

```
ENV ADMIN_USERNAME=admin
ENV ADMIN_PASSWORD=12345678
ENV ADMIN_API_KEY=admin.apikey
```

But I guess in the current release with the addition of the `owner` role we should be using `owner` instead in the Quickstart too, any thoughts on this @frascuchon?

frascuchon replied:
The admin role still works with the quickstart image, so I would say to align with the rest of the tutorials.

alvarobartt commented on docs/_source/getting_started/installation/deployments/docker-quickstart.md:
```diff
@@ -1,9 +1,8 @@
 
-Argilla is composed of a Python Server with Elasticsearch as the database layer, and a Python Client to create and
-manage datasets.
+Argilla is composed of a Python Server with Elasticsearch as the database layer, and a Python Client to create and manage datasets.
```
Now we also support user and workspace management, so maybe it's nice to include this here too!

```suggestion
Argilla is composed of a Python Server with Elasticsearch as the database layer, and a Python Client to create and manage datasets, users, and workspaces.
```

alvarobartt commented on docs/_source/getting_started/installation/deployments/huggingface-spaces.md:
````diff
@@ -2,6 +2,10 @@
 Argilla nicely integrates with the Hugging Face stack (`datasets`, `transformers`, `hub`, and `setfit`), and now it can also be deployed using the Hub's Spaces.
 
+```{warning}
+**Hugging Face Spaces are reset** every 48 hours (without activity) and **do not have data retention**, so make sure to export your data after annotating or enable data retention within the spaces config.
````
```suggestion
**HuggingFace Spaces are reset** after 48 hours of inactivity and **do not have persistent storage**, so make sure to export your data after each annotation session to not lose your data.
```

alvarobartt commented on docs/_source/guides/llms/practical_guides/collect_responses.md:
```diff
@@ -81,12 +81,15 @@ If you feel that the disagreement between annotators is too high, especially for
 
 Once you have unified your responses, you will have a dataset that"s ready for [fine-tuning](fine_tune.ipynb). Remember to save your unified dataset following one of the methods explained in [Export a Feedback dataset](export_dataset.ipynb).
 
+### Unifying RatingQuestions
```
Maybe we could name this sub-section `### Unifying` (h3) and add `#### RatingQuestion` and the rest of the questions (h4), and just define the options as bullet points

davidberenstein1957 replied:
I agree

alvarobartt commented on docs/_source/guides/llms/practical_guides/fine_tune_others.md:
```diff
@@ -0,0 +1,194 @@
+from argilla import FeedbackDataset, TrainingTaskMapingForTextClassification
+
+label_unification = ...
+dataset = FeedbackDataset(...)
```
Same thing here, one cannot just init a `FeedbackDataset` and expect it to have records, so I'd suggest to showcase either `from_argilla` or `from_huggingface`

alvarobartt commented on docs/_source/guides/llms/practical_guides/fine_tune_others.md:
```diff
@@ -0,0 +1,194 @@
+    strategy="majority"
+)
+
+training_task_mapping = rg.TrainingTaskMapingForTextClassification(
```
Also the naming for `TrainingTaskMappingForTextClassification` seems too long to me, maybe we can shorten it as `TrainingMapping(task="text-classification", ...)` or something similar? WDYT?

davidberenstein1957 replied:
Yes I agree. I was a bit unsure about the naming. Similar to Unfications, Strategies etc.

davidberenstein1957 replied:
I think I prefer to have the mapping more explicit because each task might require different field question attributes.

alvarobartt replied:
Hmm and what about providing those via `__init__` in a generic way e.g. `TrainMapping(fields=..., *args, **kwargs)`, and have some classmethods to automatically build the class with a certain configuration? e.g. `TrainMapping.for_text_classification()`

davidberenstein1957 replied:
Sure, sounds good. You mean `TrainMapping(task=..., *args, **kwargs)` right?

alvarobartt commented on src/argilla/client/feedback/schemas.py:
```diff
@@ -451,3 +499,263 @@ class FeedbackDatasetConfig(BaseModel):
         smart_union = True
+
+
+class RatingQuestionStrategy(Enum):
```
These are not schemas as is, so I'd suggest to move this to another module under `feedback` or even to create a separate directory for `training` or `training_utils`

davidberenstein1957 replied:
I was unsure about the place for this but for me it seemed more generic, not specific to training. I will create a maybe a `unification` directory?

alvarobartt replied:
Maybe a module named `unification.py` is enough, as we'll include there the different unification classes and that's it, right?

davidberenstein1957 replied:
Sure

alvarobartt commented on src/argilla/client/feedback/schemas.py:
```diff
@@ -293,6 +317,18 @@ def update_settings(cls, values: Dict[str, Any]) -> Dict[str, Any]:
         values["settings"]["options"] = [{"value": value} for value in values.get("values")]
         return values
 
+    @property
+    def __all_labels__(self):
+        return [entry["value"] for entry in self.settings["options"]]
+
+    @property
+    def __label2id__(self):
+        return {label: idx for idx, label in enumerate(self.__all_labels__)}
+
+    @property
+    def __id2label__(self):
+        return {idx: label for idx, label in enumerate(self.__all_labels__)}
+
```
`FieldSchema` has no labels and never will, so this should be removed

```suggestion
```

alvarobartt commented on src/argilla/client/feedback/schemas.py:
```diff
@@ -319,6 +355,18 @@ def update_settings(cls, values: Dict[str, Any]) -> Dict[str, Any]:
         )  # `None` is a possible value, which means all labels are visible
         return values
 
+    @property
+    def __all_labels__(self):
+        return [entry["value"] for entry in self.settings["options"]]
+
+    @property
+    def __label2id__(self):
+        return {label: idx for idx, label in enumerate(self.__all_labels__)}
+
+    @property
+    def __id2label__(self):
+        return {idx: label for idx, label in enumerate(self.__all_labels__)}
+
```
Since this will just apply to `RatingQuestion` and `_LabelQuestion` shouldn't we move it there? Otherwise those properties will fail with the rest of the questions

alvarobartt commented on src/argilla/client/feedback/training/base.py:
```diff
@@ -0,0 +1,288 @@
+_________________________________________________________________
+`trainer.train(output_dir)` to train to start training. `output_dir` is the directory to save the model automatically.
+`trainer.predict(text, as_argilla_records=True)` to make predictions.
+`trainer.save(output_dir)` to save the model manually."""
```
In these blocks we could use `textwrap` which is a Python built-in tool to avoid that weird formatting from happening, while having the same output, without the `\t` 😄 

```suggestion
        return textwrap.dedent(
            f"""\
            ArgillaBaseTrainer info:
            _________________________________________________________________
            These baseline params are fixed:
                dataset: {self._dataset}
                task: {self._training_task_mapping}
                train_size: {self._train_size}
                seed: {self._seed}
            
            
            {self._trainer.__class__} info:
            _________________________________________________________________
            The parameters are configurable via `trainer.update_config()`:
                {self._trainer}
            
            Using the trainer:
            _________________________________________________________________
            `trainer.train(output_dir)` to train to start training. `output_dir` is the directory to save the model automatically.
            `trainer.predict(text, as_argilla_records=True)` to make predictions.
            `trainer.save(output_dir)` to save the model manually.
            """
        )
```

davidberenstein1957 replied:
Thanks! Did not know this one.

alvarobartt commented on src/argilla/client/feedback/dataset.py:
As mentioned previously I wouldn't add the `training_utils` as part of the `FeedbackDataset` itself, but as separate functions, otherwise we'll end up having a monster class hard to maintain, the more detached the better IMO

davidberenstein1957 replied:
In terms of? I just wanted to add the prepare for training method and the rest is separated right?

alvarobartt reviewed (changes requested):
Hi @davidberenstein1957 good works! Just reviewed some files because this PR seems a bit too big, maybe next time we can just make these ones lighter even though I know it's hard to! I'll play around with this as soon as I've time too, but overall looks great 🎉 Thanks!

davidberenstein1957 committed:
docs: updated Labeling Feedback snippet

davidberenstein1957 committed:
Update README.md

Co-authored-by: Alvaro Bartolome <alvaro@argilla.io>

davidberenstein1957 committed:
Apply suggestions from code review alvarobart

Co-authored-by: Alvaro Bartolome <alvaro@argilla.io>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

davidberenstein1957 committed:
Apply suggestions from code review

Co-authored-by: Alvaro Bartolome <alvaro@argilla.io>

davidberenstein1957 committed:
docs: resolved https://github.com/argilla-io/argilla/pull/3151/files#r1236431524

davidberenstein1957 commented on src/argilla/client/feedback/schemas.py:
```diff
@@ -49,6 +53,23 @@ class ValueSchema(BaseModel):
+        >>> # or use a dict
+        >>> value = {"value": "Yes", "strategy": "majority"}
+    """
+
```
not re-use this and create a new one

davidberenstein1957 commented on src/argilla/client/feedback/schemas.py:
```diff
@@ -129,6 +152,7 @@ def responses_must_be_a_list(cls, v: Optional[Union[ResponseSchema, List[Respons
 
     class Config:
         extra = Extra.ignore
+        fields = {"unified_responses": {"exclude": True}}
```
write warning about unificaiton

davidberenstein1957 commented on src/argilla/client/feedback/schemas.py:
```diff
@@ -118,6 +140,7 @@ class FeedbackRecord(BaseModel):
     metadata: Optional[Dict[str, Any]] = None
     responses: Optional[Union[ResponseSchema, List[ResponseSchema]]] = None
     external_id: Optional[str] = None
+    unified_responses: Optional[Dict[str, List[UnificatiedValueSchema]]] = {}
```
__unified_responses

davidberenstein1957 commented on src/argilla/client/models.py:
```diff
@@ -37,14 +37,27 @@
+        span_marker: SpanMarker Tom Aarsen library
+        spark-nlp: Spark NLP John Snow Labs library
+        openai: OpenAI LLMs
+        autotrain: AutoTrain Transformers library
```
remove it here too

davidberenstein1957 committed:
fix: lowered typer version for better spaCy compatibility

davidberenstein1957 committed:
fix: updated deprecated logger.warn to warning

davidberenstein1957 committed:
chore: renamed TrainingTaskMappingForTokenClassification
chore: added label properties to RatingQuestion
chore: made unified_responses private attribute
fix: remove deprecated _logger methods

davidberenstein1957 committed:
fix: remove deprecated .warn(

davidberenstein1957 committed:
chore: replaced unification with implicit usage

davidberenstein1957 committed:
chore: added documentation updates

davidberenstein1957 committed:
docs: updated documentation w.r.t. unifcation strategies

davidberenstein1957 committed:
docs: added reference to spaces
docs: simplified unification workflows

davidberenstein1957 committed:
docs: updated unificaiton references

davidberenstein1957 committed:
Merge branch 'develop' into feature/prepare-for-training-feedbacktask

davidberenstein1957 committed:
docs: changed rg for import

davidberenstein1957 committed:
fix: resolved failing pytests

davidberenstein1957 committed:
chore: updated import structure argilla.feedback

davidberenstein1957 committed:
chore: added a feedback module in root for rerouting imports

alvarobartt commented on src/argilla/feedback/__init__.py:
I think that what we discussed is that we should move the complete `argilla/client/feedback/...` into `argilla/feedback/...`, not just the imports, can you confirm @frascuchon?

alvarobartt commented on tests/training/test_autotrain.py:
Is there any point on keeping this? I think we can remove it and recover it later from previous releases, as having a fully commented file doesn't make much sense IMO

alvarobartt commented on tests/training/test_base.py:
```diff
@@ -54,7 +54,7 @@ def _init_trainer(_framework) -> None:
 
     if framework == "setfit":
-        with pytest.raises(NotImplementedError, match=f"{framework} only supports `TextClassification` tasks."):
+        with pytest.raises(NotImplementedError):
```
I would keep the `match` here to make sure we're catching the correct exception 😄

alvarobartt commented on src/argilla/training/base.py:
```diff
@@ -152,25 +150,27 @@ def __init__(
+        #         settings=self._settings,
+        #         seed=self._seed,
+        #         model=self.model,
+        #     )
```
Same thing here, if some code is not working I would just remove it rather than keeping it in the code, we can easily recover this from previous releases

alvarobartt commented on src/argilla/training/base.py:
```diff
@@ -320,6 +316,14 @@ def __init__(
+        else:
+            self._label_list = None
+            self._label2id = None
+            self._id2label = None
```
```suggestion
        self._label2id = self._settings.label2id or None
        self._id2label = self._settings.id2label or None
        self._label_list = list(self._label2id.keys()) or None
```

davidberenstein1957 replied:
This wouldn't work because settings could be None

alvarobartt commented on src/argilla/training/openai.py:
```diff
@@ -18,6 +18,7 @@
 
 import numpy as np
 
+from argilla import TextClassificationRecord, TokenClassificationRecord
```
I would import this from their respective module, not directly from `argilla`.

alvarobartt commented on src/argilla/client/feedback/unification.py:
```diff
@@ -0,0 +1,334 @@
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
```
```suggestion
```

alvarobartt commented on src/argilla/client/feedback/training/frameworks/span_marker.py:
```diff
@@ -0,0 +1,60 @@
+
+from datasets import DatasetDict
+
+import argilla as rg
```
Same thing here, let's avoid using `import argilla` directly in the client code, we should just use those imports when being an user, internally IMO we shouldn't

davidberenstein1957 committed:
chore: updated code for new import structure

alvarobartt commented on src/argilla/client/feedback/training/frameworks/spacy.py:
```diff
@@ -0,0 +1,120 @@
+
+from typing import Optional
+
+import argilla as rg
```
Here too!

alvarobartt commented on src/argilla/client/feedback/training/frameworks/spacy.py:
```diff
@@ -0,0 +1,120 @@
+                codes in `spaCy` at https://spacy.io/usage/models#languages. Defaults to None.
+            gpu_id:
+                the GPU ID to use. Defaults to -1, which means that the CPU will be used by default.
+                GPU IDs start in 0, which stands for the default GPU in the system, if available.
```
IMO we should short the args as we define those in the `__init__` e.g. I would keep the inherited ones first, then the ones we're defining in this specific trainer

alvarobartt commented on src/argilla/client/feedback/training/frameworks/setfit.py:
```diff
@@ -0,0 +1,50 @@
+        ArgillaTransformersTrainer.__init__(self, *args, **kwargs)
+
+        if self._record_class is not TextClassificationRecord:
+            raise NotImplementedError("SetFit only supports.")
```
```suggestion
            raise NotImplementedError("SetFit only supports the `TextClassification` task.")
```

alvarobartt commented on src/argilla/client/feedback/training/frameworks/setfit.py:
```diff
@@ -0,0 +1,50 @@
+
+import logging
+
+from argilla import TextClassificationRecord
```
Same here

alvarobartt commented on src/argilla/client/feedback/training/frameworks/openai.py:
```diff
@@ -0,0 +1,47 @@
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
```
```suggestion
```

alvarobartt commented on src/argilla/client/feedback/training/frameworks/autotrain_advanced.py:
```diff
@@ -0,0 +1,79 @@
+                if n_classes > 2:
+                    self.task = "text_multi_class_classification"
+                else:
+                    self.task = "text_binary_classification"
```
```suggestion
                self.task = "text_multi_class_classification" if n_classes > 2 else "text_binary_classification"
```

alvarobartt commented on src/argilla/client/feedback/training/frameworks/autotrain_advanced.py:
```diff
@@ -0,0 +1,79 @@
+
+        if self._record_class == rg.TextClassificationRecord:
+            if self._multi_label:
+                raise NotImplementedError("rg.TextClassificaiton multi_label is not supported by `autotrain-advanced`.")
```
```suggestion
                raise NotImplementedError("rg.TextClassification` with `multi_label=True` is not supported by `autotrain-advanced`.")
```

alvarobartt commented on src/argilla/client/feedback/training/frameworks/autotrain_advanced.py:
```diff
@@ -0,0 +1,79 @@
+        else:
+            raise NotImplementedError(
+                "rg.Text2TextRecord and rg.TokenClassificationRecord is not supported by `autotrain-advanced`."
+            )
```
```suggestion
        else:
            raise NotImplementedError(
                "`rg.Text2Text` and `rg.TokenClassification` tasks are not supported by `autotrain-advanced`."
            )
```

alvarobartt commented on src/argilla/client/feedback/training/frameworks/autotrain_advanced.py:
```diff
@@ -0,0 +1,79 @@
+
+from datasets import DatasetDict
+
+import argilla as rg
```
Same here!

alvarobartt commented on src/argilla/client/feedback/training/frameworks/autotrain_advanced.py:
```diff
@@ -0,0 +1,79 @@
+
+class ArgillaAutoTrainTrainer(ArgillaAutoTrainTrainerV1, ArgillaTrainerSkeleton):
+    def __init__(self, *args, **kwargs):
+        raise NotImplementedError("ArgillaAutoTrainTrainer is not supported by `autotrain-advanced`.")
```
If it's not supported I guess we should just remove it, we can add it back at any point, but there's no point on releasing `NotImplementedError` stuff IMO :/

davidberenstein1957 replied:
I agree, but I don't want to remove all code, because a huge part can be re-used but I want to do that when it is more stable.

alvarobartt commented on src/argilla/client/feedback/training/base.py:
```diff
@@ -0,0 +1,292 @@
+        #         prepared_data=self._prepared_data,
+        #         seed=self._seed,
+        #         model=self._model,
+        #     )
```
Same here, if not supported / not working, lets remove it!

alvarobartt commented on src/argilla/client/feedback/training/base.py:
```diff
@@ -0,0 +1,292 @@
+    def __init__(
+        self,
+        dataset: "rg.FeedbackDataset",
+        training_task_mapping: TrainingTaskMappingForTextClassification,
```
As we're in the `ArgillaTrainer` already, maybe it makes more sense to just name this `task_mapping`?

```suggestion
        task_mapping: TrainingTaskMappingForTextClassification,
```

alvarobartt commented on src/argilla/client/feedback/training/base.py:
```diff
@@ -0,0 +1,292 @@
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, List, Optional, Union
+
+import argilla as rg
```
Same here! Sorry for the spam :(

alvarobartt commented on src/argilla/client/feedback/schemas.py:
```diff
@@ -288,6 +293,18 @@ class RatingQuestion(QuestionSchema):
     settings: Dict[str, Any] = Field({"type": "rating"}, allow_mutation=False)
     values: List[int] = Field(unique_items=True, min_items=2)
 
+    @property
+    def __all_labels__(self):
+        return [entry["value"] for entry in self.settings["options"]]
+
+    @property
+    def __label2id__(self):
+        return {label: idx for idx, label in enumerate(self.__all_labels__)}
+
+    @property
+    def __id2label__(self):
+        return {idx: label for idx, label in enumerate(self.__all_labels__)}
+
```
We don't have these in the `Rating`, right?

davidberenstein1957 replied:
Officially not, but we just map rating:rating, for re-usable logic across the different Questions w.r.t. training module implementation

alvarobartt commented on src/argilla/training/setfit.py:
```diff
@@ -36,6 +37,10 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
+
+        if self._record_class is not TextClassificationRecord:
+            raise NotImplementedError("SetFit only supports.")
```
```suggestion
            raise NotImplementedError("SetFit only supports `rg.TextClassification` tasks.")
```

alvarobartt commented on src/argilla/client/datasets.py:
```diff
@@ -449,16 +452,13 @@ def prepare_for_training(
             Framework.SPAN_MARKER,
             Framework.PEFT,
-            Framework.AUTOTRAIN,
+            # Framework.AUTOTRAIN,
```
Same here, to be removed if not working

davidberenstein1957 replied:
```suggestion

```

alvarobartt commented on src/argilla/client/feedback/__init__.py:
Maybe here we could just add everything "importable" inside `argilla/client/feedback`

alvarobartt commented on src/argilla/client/feedback/dataset.py:
```diff
@@ -967,3 +990,91 @@ def from_huggingface(cls, repo_id: str, *args: Any, **kwargs: Any) -> "FeedbackD
+        self,
+        question: Union[str, LabelQuestion, MultiLabelQuestion, RatingQuestion],
+        strategy: Union[str, LabelQuestionStrategy, MultiLabelQuestionStrategy, RatingQuestionStrategy],
+    ):
```
```suggestion
    ) -> None:
```

alvarobartt commented on src/argilla/training/spacy.py:
Here we'll probably have some merge conflicts with #3256

alvarobartt commented on src/argilla/training/span_marker.py:
```diff
@@ -27,7 +27,7 @@ class ArgillaSpanMarkerTrainer(ArgillaTrainerSkeleton):
     _logger.setLevel(logging.INFO)
 
-    require_version("span_marker")
+    require_version("span_marker>=1.2")
     require_version("transformers>=4.19.0")  # <- required for span_marker evaluation
```
We should add a `require_version("datasets")` somewhere here, right?

alvarobartt commented on tests/client/feedback/test_dataset.py:
```diff
@@ -436,7 +440,6 @@ def test_push_to_huggingface_and_from_huggingface(
     assert len(dataset_from_huggingface.questions) == len(dataset.questions)
     assert all(original_question in dataset_from_huggingface.questions for original_question in dataset.questions)
-
     for hf_record, record in zip(dataset_from_huggingface.records, dataset.records):
```
Can we have that line-break back? I was basically to split the asserts to see the separate portions of the test better 😄

alvarobartt commented on tests/training/test_openai.py:
```diff
@@ -37,14 +36,12 @@ def test_setfit_train(dataset_text_classification):
-    with pytest.raises(
-        NotImplementedError, match=f"{Framework.OPENAI} does not support multi-label TextClassification tasks."
-    ):
+    with pytest.raises(NotImplementedError):
```
Same thing here with keeping the `match`

alvarobartt commented on tests/client/feedback/training/test_trainer.py:
```diff
@@ -0,0 +1,97 @@
+        Framework("span_marker"),
+        Framework("setfit"),
+        Framework("peft"),
+        # Framework("openai"),
```
If we're not running the tests for those, just remove them instead of leaving the commented line there IMO

alvarobartt commented on tests/client/feedback/training/test_schemas.py:
```diff
@@ -0,0 +1,232 @@
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
```
```suggestion
```

alvarobartt commented on pyproject.toml:
```diff
@@ -42,7 +42,7 @@ dependencies = [
     "rich <= 13.0.1",
     # for CLI
-    "typer >= 0.6.0, < 1.0.0",
+    "typer >= 0.6.0, < 0.8.0",
```
Is there any reason for this pin? If so, could you add an in-line comment to get to know in case we decide to upgrade at some point? Thanks 😄

davidberenstein1957 replied:
```suggestion
    "typer >= 0.6.0, < 0.8.0", # spaCy only supports typer<0.8.0
```

alvarobartt reviewed (changes requested):
LGTM, just some comments of some things we should look into before merging, but so far so good! Didn't review the documentation this time, but I will for sure before merging, thanks 😄

alvarobartt commented on README.md:
```diff
@@ -85,7 +85,7 @@ docker run -d --name argilla -p 6900:6900 argilla/argilla-quickstart:latest
 <hr>
 <h3><a href="https://argilla.io/blog/launching-argilla-huggingface-hub/">Deploy on Hugging Face Hub</a></h3>
-
+HuggingFace Spaces don't have persistent storage and will automatically reset after 48 hours of inactivity, which means all the data within the Space will be lost. You can enable this via the Hugging Face Spaces settings.
```
This was no longer true 😄 Feel free to update this in other documentation files and curate the wording if you feel like, thanks!

```suggestion
HuggingFace Spaces now have persistent storage, and is supported from Argilla 1.11.0 onwards, but you will need to manually activate it via the HuggingFace Space settings. Otherwise, unless you're on a paid space upgrade, after 48 hours of inactivity the space will be shut off and you will lose all the data. To avoid losing data, we highly recommend using the persistent storage layer offered by HuggingFace.
```

davidberenstein1957 committed:
chore: removed direct argilla import
chore: added correct import structure tests
chore: cleaned up code

davidberenstein1957 committed:
Apply suggestions from code review

thanks @alvarobartt

Co-authored-by: Alvaro Bartolome <alvaro@argilla.io>

davidberenstein1957 committed:
chore: re-introduced match statements pytest

davidberenstein1957 commented on README.md:
```diff
@@ -85,7 +85,7 @@ docker run -d --name argilla -p 6900:6900 argilla/argilla-quickstart:latest
 <hr>
 <h3><a href="https://argilla.io/blog/launching-argilla-huggingface-hub/">Deploy on Hugging Face Hub</a></h3>
-
+HuggingFace Spaces now have persistent storage, and is supported from Argilla 1.11.0 onwards, but you will need to manually activate it via the HuggingFace Space settings. Otherwise, unless you're on a paid space upgrade, after 48 hours of inactivity the space will be shut off and you will lose all the data. To avoid losing data, we highly recommend using the persistent storage layer offered by H ...
```
```suggestion
HuggingFace Spaces now have persistent storage and this is supported from Argilla 1.11.0 onwards, but you will need to manually activate it via the HuggingFace Spaces settings. Otherwise, unless you're on a paid space upgrade, after 48 hours of inactivity the space will be shut off and you will lose all the data. To avoid losing data, we highly recommend using the persistent storage layer offered by HuggingFace.
```

davidberenstein1957 commented on docs/_source/_common/tabs/argilla_install.md:
````diff
@@ -15,6 +15,9 @@ wget -O docker-compose.yaml https://raw.githubusercontent.com/argilla-io/argilla
 
 :::{tab-item} Hugging Face Spaces
+```{warning}
+HuggingFace Spaces don't have persistent storage and will automatically reset after 48 hours of inactivity, which means all the data within the Space will be lost. You can enable this via the Hugging Face Spaces settings.
````
```suggestion
HuggingFace Spaces now have persistent storage and this is supported from Argilla 1.11.0 onwards, but you will need to manually activate it via the HuggingFace Spaces settings. Otherwise, unless you're on a paid space upgrade, after 48 hours of inactivity the space will be shut off and you will lose all the data. To avoid losing data, we highly recommend using the persistent storage layer offered by HuggingFace.
```

davidberenstein1957 commented on docs/_source/getting_started/installation/deployments/huggingface-spaces.md:
````diff
@@ -2,6 +2,10 @@
 Argilla nicely integrates with the Hugging Face stack (`datasets`, `transformers`, `hub`, and `setfit`), and now it can also be deployed using the Hub's Spaces.
 
+```{warning}
+**HuggingFace Spaces are reset** after 48 hours of inactivity and **do not have persistent storage**, so make sure to export your data after each annotation session to not lose your data. You can enable this via the Hugging Face Spaces settings.
````
```suggestion
HuggingFace Spaces now have persistent storage and this is supported from Argilla 1.11.0 onwards, but you will need to manually activate it via the HuggingFace Spaces settings. Otherwise, unless you're on a paid space upgrade, after 48 hours of inactivity the space will be shut off and you will lose all the data. To avoid losing data, we highly recommend using the persistent storage layer offered by HuggingFace.
```

davidberenstein1957 commented on docs/_source/getting_started/quickstart_installation.ipynb:
```diff
@@ -28,6 +28,7 @@
     "\n",
     "[![deploy on spaces](https://huggingface.co/datasets/huggingface/badges/raw/main/deploy-to-spaces-lg.svg)](https://huggingface.co/new-space?template=argilla/argilla-template-space)\n",
     "\n",
+    "Note that, Hugging Face spaces don't have data persistent and reset after 48 hour of inactivity. You can enable data persistence via the Hugging Face Spaces settings.\n",
```
```suggestion
    "HuggingFace Spaces now have persistent storage and this is supported from Argilla 1.11.0 onwards, but you will need to manually activate it via the HuggingFace Spaces settings. Otherwise, unless you're on a paid space upgrade, after 48 hours of inactivity the space will be shut off and you will lose all the data. To avoid losing data, we highly recommend using the persistent storage layer offered by HuggingFace.\n",
```

davidberenstein1957 committed:
Apply suggestions from code review regarding Hugging Face Spaces
